presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -641,103 +639,137 @@ public Duration getTotalCpuTime()
-                        completedStages.add(stage.getStageId());
+                                .processScheduleResults(stage.getState(), result.getNewTasks());
+                        schedulerStats.getSplitsScheduledPerIteration().add(result.getSplitsScheduled());
+                        if (result.getBlockedReason().isPresent()) {
+                            switch (result.getBlockedReason().get()) {
+                                case WRITER_SCALING:
+                                    // no-op
+                                    break;
+                                case WAITING_FOR_SOURCE:
+                                    schedulerStats.getWaitingForSource().update(1);
+                                    break;
+                                case SPLIT_QUEUES_FULL:
+                                    schedulerStats.getSplitQueuesFull().update(1);
+                                    break;
+                                case MIXED_SPLIT_QUEUES_FULL_AND_WAITING_FOR_SOURCE:
+                                    schedulerStats.getMixedSplitQueuesFullAndWaitingForSource().update(1);
+                                    break;
+                                case NO_ACTIVE_DRIVER_GROUP:
+                                    schedulerStats.getNoActiveDriverGroup().update(1);
+                                    break;
+                                default:
+                                    throw new UnsupportedOperationException("Unknown blocked reason: " + result.getBlockedReason().get());
+                            }
+                        }
                     }
-                }
 
-                // wait for a state change and then schedule again
-                if (!blockedStages.isEmpty()) {
-                    try (TimeStat.BlockTimer timer = schedulerStats.getSleepTime().time()) {
-                        tryGetFutureValue(whenAnyComplete(blockedStages), 1, SECONDS);
+                    // make sure to update stage linkage at least once per loop to catch async state changes (e.g., partial cancel)
+                    boolean stageFinishedExecution = false;
+                    for (SqlStageExecution stage : scheduledStages) {
+                        if (!completedStages.contains(stage.getStageId()) && stage.getState().isDone()) {
+                            stageLinkages.get(stage.getStageId())
+                                    .processScheduleResults(stage.getState(), ImmutableSet.of());
+                            completedStages.add(stage.getStageId());
+                            stageFinishedExecution = true;
+                        }
                     }
-                    for (ListenableFuture<?> blockedStage : blockedStages) {
-                        blockedStage.cancel(true);
+
+                    // if any stage has just finished execution try to pull more sections for scheduling
+                    if (stageFinishedExecution) {
```

Comment:
why?  Don't we need to wait for the whole section to be scheduled anyway? That should be caught by the while condition.

Reply:
Let's say that a section is blocked by a stage. The stage has finished scheduling, what breaks the `while` loop. The method tries to pull more sections to schedule, but the stage that has finished scheduling hasn't finished the execution yet. So the new section won't be pulled in. But, since we are already checking stages for competition - we can also try to pick more sections at the moment when a stage actually finishes execution.